cycloneDX doc (#3959)
conan-io/docs: reference/tools/sbom.rst

install the dependencies from Conan. This can be very useful for generating SBOMs for different versions of our
dependencies. Note that this time we're saving the SBOM in the generators folder, so that the user installing the dependencies
has easy access to the SBOM.

.. code-block:: python

    import json
    import os
    from conan.api.output import ConanOutput
    from conan.tools.sbom.cyclonedx import cyclonedx_1_4

    def post_generate(conanfile, **kwargs):
        sbom_cyclonedx_1_4 = cyclonedx_1_4(conanfile.subgraph)
        generators_folder = conanfile.generators_folder
        file_name = "sbom.cdx.json"
        os.mkdir(os.path.join(generators_folder, "sbom"))
        with open(os.path.join(generators_folder, "sbom", file_name), 'w') as f:
            json.dump(sbom_cyclonedx_1_4, f, indent=4)
        ConanOutput().success(f"CYCLONEDX CREATED - {conanfile.generators_folder}")

Both hooks can coexist in such a way that we can generate the SBOMs for our application and our dependencies separately.
This can greatly assist us in conducting continuous analysis of our development process and ensuring software quality.

Conan
^^^^^

> Instead of using a standard, we can take a Conan-based approach. Thanks to the ``conanfile.subgraph.serialize()``

Reviewer: ```suggestion
Instead of using a standard, we can take a "Conan-based approach". Thanks to the ``conanfile.subgraph.serialize()``
```

        file_name = "sbom.cdx.json"
        with open(os.path.join(metadata_folder, file_name), 'w') as f:
            json.dump(sbom_cyclonedx_1_4, f, indent=4)
        ConanOutput().success(f"CYCLONEDX CREATED - {conanfile.package_metadata_folder}")

In the second example, we generate our SBOM after the generate method. This allows us to create the SBOMs when we
install the dependencies from Conan. This can be very useful for generating SBOMs for different versions of our
dependencies.

.. code-block:: python

    import json
    import os
    from conan.errors import ConanException
    from conan.api.output import ConanOutput
    from conan.tools.sbom.cyclonedx import cyclonedx_1_4
    def post_generate(conanfile):
        sbom_cyclonedx_1_4 = cyclonedx_1_4(conanfile.subgraph)
        generators_folder = conanfile.generators_folder
        file_name = "sbom.cdx.json"
        os.mkdir(os.path.join(generators_folder, "sbom"))
        with open(os.path.join(generators_folder, "sbom", file_name), 'w') as f:
            json.dump(sbom_cyclonedx_1_4, f, indent=4)
        ConanOutput().success(f"CYCLONEDX CREATED - {conanfile.generators_folder}")

> 

Reviewer 2: Lets add a small ``reference`` section for the ``cyclonedx_1_4()`` function.

an SBOM provides transparency about what is contained "inside" an application or software system.

Conan allows you to generate SBOMs natively by using the resolved dependency graph.
This way, you can create the SBOM for your program at the same time you build it.

For now, **this feature is in an experimental state, which means that the interface, functionality or generated
files may change in the future**. Additionally, it currently only supports CycloneDX version 1.4.
If you need a different standard, another version, or if you encounter any potential improvements,
please feel free to open an issue on our `GitHub <https://github.com/conan-io/conan/issues>`_ .
We would be delighted to hear your feedback!

CycloneDX
^^^^^^^^^

CycloneDX is one of the most widely used standards for SBOMs, supported by the OWASP Foundation.

Using this feature is as simple as implementing a :ref:`hook <reference_extensions_hooks>` in your client,
which uses this tool to create the SBOM and simply stores it in the appropriate location.

Let's look at two examples:

In the first one, we have the case where we want to generate the SBOM at the moment we create our app, after the

> package method. This is very useful for keeping track of the components and dependencies of our software.

Reviewer 3: ```suggestion
package method. This is very useful for keeping track of the components and dependencies of that went into building our software.
```

In the example, we save it in the metadata folder to keep our project organized.

.. code-block:: python

    import json
    import os
    from conan.api.output import ConanOutput
    from conan.tools.sbom.cyclonedx import cyclonedx_1_4

>     def post_package(conanfile):

Reviewer 3: ```suggestion
    def post_package(conanfile, **kwargs):
```

Most hook examples should have this so that in the future we can more easily update their signature if it's ever needed

        sbom_cyclonedx_1_4 = cyclonedx_1_4(conanfile.subgraph)
        metadata_folder = conanfile.package_metadata_folder
        file_name = "sbom.cdx.json"
        with open(os.path.join(metadata_folder, file_name), 'w') as f:
            json.dump(sbom_cyclonedx_1_4, f, indent=4)
        ConanOutput().success(f"CYCLONEDX CREATED - {conanfile.package_metadata_folder}")

In the second example, we generate our SBOM after the generate method. This allows us to create the SBOMs when we
install the dependencies from Conan. This can be very useful for generating SBOMs for different versions of our
dependencies.

.. code-block:: python

    import json
    import os
    from conan.api.output import ConanOutput
    from conan.tools.sbom.cyclonedx import cyclonedx_1_4

>     def post_generate(conanfile):

Reviewer 3: ```suggestion
    def post_generate(conanfile, **kwargs):
```

        sbom_cyclonedx_1_4 = cyclonedx_1_4(conanfile.subgraph)
        generators_folder = conanfile.generators_folder
        file_name = "sbom.cdx.json"
        os.mkdir(os.path.join(generators_folder, "sbom"))
        with open(os.path.join(generators_folder, "sbom", file_name), 'w') as f:
            json.dump(sbom_cyclonedx_1_4, f, indent=4)

>         ConanOutput().success(f"CYCLONEDX CREATED - {conanfile.generators_folder}")

Reviewer 3: Maybe a comment that both hooks are outputting their sbom in different places?

Both hooks can coexist in such a way that we can generate the SBOMs for our application and our dependencies separately.
This can greatly assist us in conducting continuous analysis of our development process and ensuring software quality.

Conan
^^^^^

Instead of using a standard, we can take a Conan-based approach. Thanks to the ``conanfile.subgraph.serialize()``
function, we can directly obtain information about the dependencies of our program.

> In the following example, we can see a hook that generates this simplified SBOM.

Reviewer 3: ```suggestion
In the following example, we can see a hook that generates a simplified SBOM
consisting of the serialization of the subgraph, which includes all data Conan has
about the specific dependencies.
```

or something like this, as I don't think either method is too documented right now

.. code-block:: python

    import json
    import os
    from conan.api.output import ConanOutput

>     def post_package(conanfile):

Reviewer 3: ```suggestion
    def post_package(conanfile, **kwargs):
```
